### PR TITLE
Implementing client certificate based authentication.

### PR DESCRIPTION
**Pull request by rushmin**

### CODE REVIEW

**Comment by omindu** [comment]
Can I know from where does the `WSO2-Identity-User` header is getting set ?

**Comment by rushmin** [comment]
It's custom header. The client should send it as an HTTP header. May be we can come up with a better name. Since this is a custom header there is no standard for the name.

File: components/org.wso2.carbon.identity.auth.service/src/main/java/org/wso2/carbon/identity/auth/service/handler/HandlerManager.java [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+
+    /**
+     * Get the first priority handler after sort and filter the enabled handlers.
+     *
+     * @param identityHandlers
+     * @param isEnableHandlersOnly
+     * @return IdentityHandler
+     */
+    public <T extends IdentityHandler> T getFirstPriorityHandler(List<T> identityHandlers,
+                                                                 boolean isEnableHandlersOnly) {
```
Comment (mefarazath) [inline components/org.wso2.carbon.identity.auth.service/src/main/java/org/wso2/carbon/identity/auth/service/handler/HandlerManager.java]: I guess we are using 'isEnableHandlersOnly' flag to decide whether we only check enabled handlers or not. Can we change the variable name to 'isEnabledHandlersOnly' or 'checkEnabledHandlerOnly'? 

IMO that would give better :)

Comment (rushmin) [inline components/org.wso2.carbon.identity.auth.service/src/main/java/org/wso2/carbon/identity/auth/service/handler/HandlerManager.java]: Yes that should be a typo :-) Fixed it with the suggested name 'checkEnabledHandlerOnly'

Comment (mefarazath) [inline components/org.wso2.carbon.identity.auth.service/src/main/java/org/wso2/carbon/identity/auth/service/handler/HandlerManager.java]: great! :+1:

**Comment by omindu** [comment]
Okay got it. Thanks! Just want to understand how this works. Seems this is different than how our current X509 authenticator works. It obtains the username from the certificate CN. 

Also, what happens if the username is not there in the user store ?

**Comment by rushmin** [comment]
There are some use cases, where one application  should talk to the other application and the authenticated user is determined by the first app. (e.g, the logged in user of the first app) 

In that case treating the certificate CN as the user would not help.

A not available user is not handled in the current code. How does it work in the X509 authenticator ?

**Comment by omindu** [comment]
Okay that makes sense. I can't remember the exact outcome of the behavior of the X509 authenticator when the username is not there. @hmrajas might have an input.

**Comment by hmrajas** [comment]
HI All,
In current X509 certificate authenticator implementation it will take the CN as username if only enabled authenticator as first step otherwise it will chose authenticated user username as the username add the certificate to user. if the username not there in the userstore in first step simply it will return exception[1].


[1] "Caused by: org.wso2.carbon.user.core.UserStoreException: UserNotFound: User testuserdoes not exist in: PRIMARY"

**Pull request merged**